theories/algebra/IntDiv.ec

```diff
@@ -864,9 +864,37 @@ move=> lt_pn; rewrite {1}(divz_eq i (k^n)); rewrite -addrA; congr.
 by rewrite {1}(divz_eq (i %% k^n) (k^p)) modz_dvd_pow.
 qed.
 
-lemma nosmt modz_pow2_div n p i: 0 <= i => 0 <= p <= n =>
-  (i %% 2^n) %/ 2^p = (i %/ 2^p) %% 2^(n-p).
-proof. admitted.
+lemma nosmt divzMr i a b :  
+  0 <= a => 0 <= b => i %/ (a * b) = i %/ a %/ b.
+proof.
+move=> ge0_a ge0_b.
+case (a = 0); first by smt(div0z). 
+case (b = 0) => [| neq0_b neq0_a]; first by smt(div0z).
+pose ab := a * b; move: (edivzP i ab) (divz_eq i a) (divz_eq (i %/ a) b) => [].
+rewrite mulf_eq0 neq0_b neq0_a /= => eqi_ab rngi_ab eqi_a eqia_b.
+move: (euclideU ab (i %/ ab) (i %/ a %/ b) (i %% ab) (i %/ a %% b * a + i %% a)).
+have ->: i %/ a %/ b * ab + (i %/ a %% b * a + i %% a) = i.
++ by rewrite /ab (mulrC a) mulrA addrA -mulrDl -eqia_b -eqi_a.
+rewrite -eqi_ab rngi_ab andaE /=.
+have -> //: 0 <= i %/ a %% b * a + i %% a /\ i %/ a %% b * a + i %% a < `|ab|.
+split; first by rewrite addz_ge0 1:mulr_ge0 1,3:modz_ge0.
+rewrite ger0_norm 1:mulr_ge0 // (ltr_le_trans ((b - 1) * a + a) _ ab).
```

Comment:
Would the `rewrite (ltr_le_trans ...)` allow slightly inference as an `apply: (ltr_le_trans ((b - 1) * a + a))`?